Mathlib/Probability/Kernel/Disintegration/Basic.lean

```diff
@@ -0,0 +1,158 @@
+open scoped ENNReal MeasureTheory Topology
+
+variable {α β Ω : Type*} {mα : MeasurableSpace α} {mβ : MeasurableSpace β} [MeasurableSpace Ω]
+
+/-!
+### Disintegration of measures
+
+This section provides a predicate for a kernel to disintegrate a measure.
+-/
+
+namespace MeasureTheory.Measure
+variable (ρ : Measure (α × Ω)) [IsFiniteMeasure ρ] (ρCond : Kernel α Ω)
+
+/-- A kernel `ρCond` is a conditional kernel for a measure `ρ` if it disintegrates it in the sense
+that `ρ.fst ⊗ₘ ρCond = ρ`. -/
+class IsCondKernel : Prop where
+  disintegrate' : ρ.fst ⊗ₘ ρCond = ρ
+
+variable [ρ.IsCondKernel ρCond]
+
+lemma disintegrate : ρ.fst ⊗ₘ ρCond = ρ := IsCondKernel.disintegrate'
+
+/-- Auxiliary lemma for `IsCondKernel.apply_of_ne_zero`. -/
+private lemma IsCondKernel.apply_of_ne_zero_of_measurableSet [MeasurableSingletonClass α]
+    [IsSFiniteKernel ρCond] {x : α} (hx : ρ.fst {x} ≠ 0) {s : Set Ω} (hs : MeasurableSet s) :
+    ρCond x s = (ρ.fst {x})⁻¹ * ρ ({x} ×ˢ s) := by
+  nth_rewrite 2 [← ρ.disintegrate ρCond]
+  rw [Measure.compProd_apply (measurableSet_prod.mpr (Or.inl ⟨measurableSet_singleton x, hs⟩))]
+  classical
+  have (a) : ρCond a (Prod.mk a ⁻¹' {x} ×ˢ s) = ({x} : Set α).indicator (ρCond · s) a := by
+    obtain rfl | hax := eq_or_ne a x
+    · simp only [singleton_prod, mem_singleton_iff, indicator_of_mem]
+      congr with y
+      simp
+    · simp only [singleton_prod, mem_singleton_iff, hax, not_false_eq_true, indicator_of_not_mem]
+      have : Prod.mk a ⁻¹' (Prod.mk x '' s) = ∅ := by ext y; simp [Ne.symm hax]
+      simp only [this, measure_empty]
+  simp_rw [this]
+  rw [MeasureTheory.lintegral_indicator _ (measurableSet_singleton x)]
+  simp only [Measure.restrict_singleton, lintegral_smul_measure, lintegral_dirac]
+  rw [← mul_assoc, ENNReal.inv_mul_cancel hx (measure_ne_top _ _), one_mul]
+
+/-- If the singleton `{x}` has non-zero mass for `ρ.fst`, then for all `s : Set Ω`,
+`ρCond x s = (ρ.fst {x})⁻¹ * ρ ({x} ×ˢ s)` . -/
+lemma IsCondKernel.apply_of_ne_zero [MeasurableSingletonClass α] [IsSFiniteKernel ρCond] {x : α}
+    (hx : ρ.fst {x} ≠ 0) (s : Set Ω) : ρCond x s = (ρ.fst {x})⁻¹ * ρ ({x} ×ˢ s) := by
+  have : ρCond x s = ((ρ.fst {x})⁻¹ • ρ).comap (fun (y : Ω) ↦ (x, y)) s := by
+    congr 2 with s hs
+    simp [IsCondKernel.apply_of_ne_zero_of_measurableSet _ _ hx hs,
+      (measurableEmbedding_prod_mk_left x).comap_apply]
+  simp [this, (measurableEmbedding_prod_mk_left x).comap_apply, hx]
+
+-- TODO: Can we generalise away from `MeasurableSingletonClass`, using eg `measurableAtom a` instead
+-- of `{a}`?
+lemma IsCondKernel.isMarkovKernel_iff_isSFiniteKernel [MeasurableSingletonClass α]
+    (hρ : ∀ a, ρ.fst {a} ≠ 0) : IsMarkovKernel ρCond ↔ IsSFiniteKernel ρCond := by
+  refine ⟨fun _ ↦ inferInstance, fun _ ↦ ⟨fun a ↦ ⟨?_⟩⟩⟩
+  rw [IsCondKernel.apply_of_ne_zero _ _ (hρ _), prod_univ, ← Measure.fst_apply
+    (measurableSet_singleton _), ENNReal.inv_mul_cancel (hρ _) (measure_ne_top _ _)]
```

Comment:
The hypothesis `∀ a, ρ.fst {a} ≠ 0` is very restrictive. The more useful result might be that if `ρ.fst {a} ≠ 0` then `ρCond a` is a probability measure, which can be extracted from this proof. Can you add that lemma?

The more general version of this is the `ae_markov` lemma I wrote in my first comment for kernels and its (yet to be written) corollary for measures. But let's leave that for a future PR.